test/case/ietf_interfaces/speed_duplex_copper/test.py

```diff
@@ -0,0 +1,203 @@
+#!/usr/bin/env python3
+"""
+Interface Speed Duplex (Copper)
+
+Verify that auto-negotiation results in expected speed/duplex mode.
+"""
+
+import infamy
+import subprocess
+from infamy.util import until
+
+ADVERTISE_MODES = {
+    # Values from ethtool's ETHTOOL_LINK_MODE bit positions
+    # See: https://elixir.bootlin.com/linux/latest/source/include/uapi/linux/ethtool.h
+    "10half": 0x0001,
+    "10full": 0x0002,
+    "100half": 0x0004,
+    "100full": 0x0008,
+    "1000full": 0x0020
+}
+
+def advertise_host_modes(interface, modes):
+    mask = 0
+    for mode in modes:
+        mask |= ADVERTISE_MODES[mode]
+    try:
+        subprocess.run([
+            "ethtool", "-s", interface,
+            "advertise", hex(mask)
+        ], check=True)
+    except subprocess.CalledProcessError as e:
+        raise RuntimeError(f"Failed to advertise modes via ethtool: {e}")
+
+def get_target_speed_duplex(target, interface):
+    data = target.get_data(f"/ietf-interfaces:interfaces/interface[name='{interface}']") \
+                  ["interfaces"]["interface"][interface]
+    eth = data.get("ethernet", {})
+    
+    return eth.get("speed"), eth.get("duplex")
+
+def set_target_speed_duplex(target, interface, speed, duplex):
+    target.put_config_dicts({
+            "ietf-interfaces": {
+                "interfaces": {
+                    "interface": [{
+                        "name": interface,
+                        "ethernet": {
+                            "auto-negotiation": {
+                                "enable": False
+                            },
+                            "speed": speed / 1000,
+                            "duplex": duplex
+                        }
+                    }]
+                }
+            }
+        })
+
+def set_host_speed_duplex(interface, speed, duplex):
+    try:
+        subprocess.run([
+            "ethtool", "-s", interface,
+            "speed", str(speed),
+            "duplex", duplex,
+            "autoneg", "off"
+        ], check=True)
+    except subprocess.CalledProcessError as e:
+        raise RuntimeError(f"Failed to set speed/duplex via ethtool: {e}")
+
+def verify_speed_duplex(target, interface, exp_speed, exp_duplex):
+    until(lambda: speed_duplex_present(target, interface))
+    act_speed, act_duplex = get_target_speed_duplex(target, interface)
+    if act_speed is None or act_duplex is None:
+        print(f"Could not fetch speed or duplex from target for interface {interface}")
+        test.fail()
+
+    exp_speed_gbps = exp_speed / 1000
+    if float(act_speed) != exp_speed_gbps:
+        print(f"act_speed: {act_speed}, exp_speed: {exp_speed_gbps}")
+        test.fail()
+
+    if act_duplex.lower() != exp_duplex.lower():
+        print(f"act_duplex: {act_duplex}, exp_duplex: {exp_duplex}")
+        test.fail()
+
+    print(f"Verified: {interface} is operating at {act_speed} Gbps, {act_duplex} duplex")
```

Comment:
My bad for not catching this in the first round, but I think we want to do a ping check before declaring victory here. With things like duplex mismatch, you can easily have links that are reported as up, but without bidirectional forwarding.